### PR TITLE
260/eth wrapper

### LINKED ISSUE #260: [on hold] Wrap ETH
We should allow to WRAP and unwrap ether easily...

Ideally integrated in the app. At least a link to somewhere we can wrap/unwrap.

### PR DESCRIPTION
**Pull request by W3stside**
closes #260  
ETH Wrapper module

Uses: [Wraptor](https://github.com/W3stside/wraptor)

1. updates auto on new blockheader
2. catches bad amounts/ does data validation

### CODE REVIEW

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr291--dexreact.review.gnosisdev.com

File: src/components/WrapEtherWidget/index.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import { tokenListApi } from 'api'
+import { Network } from 'types'
+
+const WrapEther: React.FC = () => {
+  const { userAddress, networkId, web3 } = useWalletConnection()
+
+  const fallBackNetworkId = networkId ? networkId : Network.Mainnet // fallback to mainnet
+  const tokens = useMemo(() => tokenListApi.getTokens(fallBackNetworkId), [fallBackNetworkId])
+
+  const contractAddress = useMemo(() => tokens.find(token => token.symbol === 'WETH')?.address, [tokens])
```
Comment (Velenir) [inline src/components/WrapEtherWidget/index.tsx]: As we only use `tokens` here, put it all into one `useMemo`

File: src/api/wallet/WalletApiImpl.ts [hunk 2]
```diff
@@ -134,7 +134,7 @@ const closeOpenWebSocketConnection = (web3: Web3): void => {
 export class WalletApiImpl implements WalletApi {
   private _listeners: ((walletInfo: WalletInfo) => void)[]
   private _provider: Provider | null
-  private _web3: Web3
+  public _web3: Web3
```
Comment (Velenir) [inline src/api/wallet/WalletApiImpl.ts]: Why `public`?

Comment (W3stside) [reply src/api/wallet/WalletApiImpl.ts]: I am returning it inside the `useWalletInfo` hook. Thoughts?

File: src/components/WrapEtherWidget/index.tsx (hunk 1 above)
Comment (W3stside) [reply src/components/WrapEtherWidget/index.tsx]: ok!

File: src/api/wallet/WalletApiImpl.ts (hunk 2 above)
Comment (Velenir) [reply src/api/wallet/WalletApiImpl.ts]: That shouldn't affect anything
Public is for `new WalletApiImpl().web3`
Returns are unrelated

Comment (W3stside) [reply src/api/wallet/WalletApiImpl.ts]: :+1:

File: src/api/wallet/WalletApiImpl.ts [hunk 3]
```diff
@@ -248,13 +247,14 @@ export class WalletApiImpl implements WalletApi {
       isConnected,
       userAddress: accounts[0],
       networkId: isConnected ? +chainId : undefined,
+      web3: this._web3,
     }
   }
 
   private async _notifyListeners(): Promise<void> {
-    await Promise.resolve()
+    const blockNumber = await this._blockNumber
```
Comment (alfetopito) [inline src/api/wallet/WalletApiImpl.ts]: No need for parenthesis when calling `this._blockNumber`?

File: src/components/WrapEtherWidget/index.tsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import { Network } from 'types'
+
+const WrapEther: React.FC = () => {
+  const { blockNumber, userAddress, networkId, web3 } = useWalletConnection()
+
+  const contractAddress = useMemo(() => {
+    const fallBackNetworkId = networkId ? networkId : Network.Mainnet // fallback to mainnet
+    const tokens = tokenListApi.getTokens(fallBackNetworkId)
+
+    return tokens.find(token => token.symbol === 'WETH')?.address
```
Comment (alfetopito) [inline src/components/WrapEtherWidget/index.tsx]: Use `getToken` function from utils?

File: src/api/wallet/WalletApiImpl.ts (hunk 3 above)
Comment (W3stside) [reply src/api/wallet/WalletApiImpl.ts]: Getter, so nope!

Comment (alfetopito) [reply src/api/wallet/WalletApiImpl.ts]: Cool, good to know!

Comment (Velenir) [inline src/api/wallet/WalletApiImpl.ts]: Try
```ts
private async _notifyListeners({blockHeader}): Promise<void> {
    const walletInfo: WalletInfo = this._getWalletInfo()
    const wInfoExtended = {...walletInfo, blockNumber: blockHeader?.number}
    this._listeners.forEach(listener => listener(wInfoExtended))
  }

File: src/api/wallet/WalletApiImpl.ts [hunk 5]
```diff
@@ -248,13 +247,14 @@ export class WalletApiImpl implements WalletApi {
...
     const walletInfo: WalletInfo = this._getWalletInfo()
-    this._listeners.forEach(listener => listener(walletInfo))
+    this._listeners.forEach(listener => listener({ ...walletInfo, blockNumber }))
```
Comment (Velenir) [inline src/api/wallet/WalletApiImpl.ts]:
```js
{ ...walletInfo, blockNumber }
```
will get recreated for each listener, construct outside the loop

Comment (W3stside) [reply src/api/wallet/WalletApiImpl.ts]: :+1:

File: src/components/WrapEtherWidget/index.tsx (hunk 4 above)
Comment (W3stside) [reply src/components/WrapEtherWidget/index.tsx]: :+1:

**Comment by W3stside** [comment]
> Cool!!
> 
> I have a few functional comments to make it more UX friendly.
> 
> * I wouldn't add a link in the navigation menu for it
> * I think it's ok and seems simpler to follow your approach of having on it's own page. I think we can keep it like that.
> * We should navigate to the page by clicking this button:
>   image
> * For the allowance, we should be coherent with the other allowances handling. Detect if WET is enabled, and if it's not, show only a "Enable WETH" button
> * Display the WETH balance automatically, on top of the unwrap input, ideally with a set max function
> * Display the ETH balance automatically, on top of the wrap input, ideally with a set max function
> * If we are implementing the wrapper ourselves, do you think is worth it to use the component. I mean, most of the logic, formatters, approve and use-max functionality and so on it's already in this project. I'm not 100% using the component would make this development simpler
> 
> Screenshot at Dec 09 19-12-49

good ideas, will implement.

in regards to using the component - there is also a hook: [`useWraptor`](https://github.com/W3stside/wraptor/blob/master/src/hooks/useWraptor.ts) that provides the logic and is exported - we can use that

File: src/api/wallet/WalletApiImpl.ts (hunk 3 above)
Comment (W3stside) [reply src/api/wallet/WalletApiImpl.ts]: cool, done. thanks!

**Comment by anxolin** [comment]
> in regards to using the component - there is also a hook: useWraptor that provides the logic and is exported - we can use that

Nicee

**Comment by W3stside** [comment]
Good point by @anxolin:

> If we are implementing the wrapper ourselves, do you think is worth it to use the component. I mean, most of the logic, formatters, approve and use-max functionality and so on it's already in this project. I'm not 100% using the component would make this development simpler

What do you guys think @alfetopito @Velenir ? Vote :+1: if we should add it, or :-1: if we shouldn't

**Comment by anxolin** [comment]
> What do you guys think @anxolin @alfetopito @Velenir ? Vote 👍 if we should add it, or 👎 if we shouldn't

Imo. I would leave this decision to you. I just want to make whatever is simpler. If you feel is not very difficult to adapt the component to our needs and use it, go for it. If you think it's easier to do it as we've done the other components, maybe in the future we can extract part of that functionality and move it to the component. You are the best to take this decision, use your best judgement :)

**Pull request closed without merging**